content/en/docs/tasks/tls/manual-rotation-of-ca-certificates.md

```diff
@@ -0,0 +1,150 @@
+    
+    You can also use [`cfssl`](/docs/concepts/cluster-administration/certificates/#cfssl).
+
+1. If your cluster is using bootstrap tokens to join nodes, update the ConfigMap `cluster-info` in the `kube-public` namespace.
+
+    ```shell
+    base64_encoded_ca="$(base64 /etc/kubernetes/pki/ca.crt)"
+
+    kubectl get cm/cluster-info --namespace kube-public -o yaml | \
+        /bin/sed "s/\(certificate-authority-data:\).*/\1 ${base64_encoded_ca}" | \
+        kubectl apply -f -
+    ```
+
+1. Update all service account tokens with the new CA certificates.
+    ```shell
+    base64_encoded_ca="$(base64 /etc/kubernetes/pki/ca.crt)"
+
+    for namespace in $(kubectl get ns --no-headers | awk '{print $1}'); do
+        for token in $(kubectl get secrets --namespace "$namespace" --field-selector type=kubernetes.io/service-account-token -o name); do
+            kubectl get $token --namespace "$namespace" -o yaml | \
+            /bin/sed "s/\(ca.crt:\).*/\1 ${base64_encoded_ca}" | \
+            kubectl apply -f -
+        done
+    done
+
+    ```
+
+1. Update certificates for user accounts.
+
+    For information about creating certificates for individual user accounts, see [Configure certificates for user accounts](/docs/setup/best-practices/certificates/#configure-certificates-for-user-accounts).
+
+    Update `client-certificate-data` and `client-key-data` respectively with the contents of certificates and private keys generated in their respective conf files, along with the kubelet client file in`/var/lib/kubelet/kubelet-client-current.pem`.
+
+1. Update `kubelet.conf`, the `kubelet` certificate, and the private key on all worker nodes as well.
+
+1. Update the `certificate-authority-data` section in all `admin.conf`, `controller-manager.conf`,  `scheduler.conf`, and `kubelet.conf` files with Base64-encoded CA certificate data:
+
+    Example:
+
+    ```shell
+    base64_encoded_ca="$(base64 /etc/kubernetes/pki/ca.crt)"
+
+    /bin/sed -i "s/\(certificate-authority-data:\).*/\1 ${base64_encoded_ca}" /path/to/conf/files
+    ```
+
+1. Follow these steps on all control plane and worker nodes.
+
+    1. Stop the kubelet service on all control plane and worker nodes.
+
+    1. Kill the containers manually on control plane nodes:
+
+        Below example assumes the container runtime in use is `Docker`.
+
+        ```shell        
+        docker ps -af --name="k8s_*" -q | xargs --no-run-if-empty docker rm -f -
+        ``` 
+
+    1. Start the kubelet service.
+
+    1. Restart any other *[aggregated api servers](/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/)* or *webhook handlers* to trust the new CA certificates.
+
+    1. Make sure the control plane comes up and logs no TLS errors.
+
+    1. Make sure container networking functions.
+
+    1. Make sure `coredns` and `kube-proxy` are updated with the new CA and function correctly after CA rotation.
```

Comment:
Maybe add a sentence stating that "If CoreDNS and KubeProxy are managed by Deployments and DaemonSets respectively, a rolling deletion of existing pods will cause new pods to be created with an auto-mounted ServiceAccount secret that includes both the old and new CA data.

Reply:
@micahhausler the amended steps I mentioned does not append CA to serviceaccounts but replace it. We will have old + new CA appended in clientCAFile or rootCAFile resp. As long as that is there we can renew all other configs including service accounts and kubeconfigs with new CA. This will help validating the new CA easily and will also reduce the impact when the control plane is restarted again after removal of old CA.

Please let me know if my understanding is correct here.

Reply:
For ServiceAccount secrets, the guide should indicate that the following transition is made:

* All service account secrets include the old CA, all pods with a kube client trust the old CA (initial state) 
* User restarts controller manager with `-–root-ca-file` pointing to file that includes both old and new CAs
  * Any new service accounts created after this point will trust both old and new CAs
* User updates all existing service account secrets `ca.crt` data field to include old CA appended by new CA
  * If any pods are restarted before new CA is used by API servers, they will pick up this update and trust both old and new CAs
* Restart all pods using in-cluster configs (ex: kube-proxy, coredns, etc) so they use the updated CA secret material 
* User restarts API servers with cert signed by new CA
  * A pod's kube client disconnects from a terminating API server that had a cert using the old CA, connects to a new API server that uses a cert from the new CA. Since the pod trusts both CAs, there is only a momentary disconnect/reconnect. 

[optional cleanup] 
* KCM can be restarted with only new CA in the `--root-ca-file`
  * Any new pods started after this point will only trust the new CA
* SA secrets can be updated to only include new CA
  * All pods using an in-cluster kubeconfig will eventually need to be restarted to pick up the new SA secret for the old CA to be completely untrusted

Reply:
After re-iterating through all steps, I am in complete agreement here. Will update the doc and send for your review shortly.

Reply:
Made changes, please review again and approve if its all good.